builtins: implement pg_get_serial_sequence

otan opened the pull request:
Encountered this one whilst importing geospatial data - we need to
ability to use pg_get_serial_sequence for importing certain dumps
for ogr2ogr.

We have no notion of saving "SERIAL" at the moment, so make some
heuristics to calculate this. Let me know if we should be doing
something else.

Release note (sql change): Add the pg_get_serial_sequence builtin.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52436)
<!-- Reviewable:end -->

dt commented:
when you say "for imports" do you mean for IMPORT?

otan commented:
yeah

arulajmani reviewed (changes requested):
Looks good in general, left some thoughts on detecting use of `SERIAL` vs default expression using a single sequence. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52436#-:-ME3NJXcCwLVdEjwcFsv:b-pwwp18)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @dt, and @otan)

___
*[pkg/sql/sequence.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52436#-ME3JuAM-Ri2WAHz58X8:-ME3JuANB5pceCLzgCip:bjvd1c0) ([raw file](https://github.com/cockroachdb/cockroach/blob/018a7a2fddbc00fda40d995e12cfa0965f708e03/pkg/sql/sequence.go#L48)):*
> ```Go
> 		if col.ColName() == columnName {
> 			// Seems like we have no way of detecting whether this was done using "SERIAL".
> 			// Guess by assuming it is SERIAL if it is only used by 1 sequence.
> ```

nit: s/used/uses/g
___
*[pkg/sql/sequence.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52436#-ME3Fvno9tmZ789AepMF:-ME3Fvno9tmZ789AepMG:bhz8xiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/018a7a2fddbc00fda40d995e12cfa0965f708e03/pkg/sql/sequence.go#L49)):*
> ```Go
> 			// Seems like we have no way of detecting whether this was done using "SERIAL".
> 			// Guess by assuming it is SERIAL if it is only used by 1 sequence.
> 			if len(col.UsesSequenceIds) == 1 {
> ```

Can we save this information on the column descriptor during `processSerialInColumnDef`? I see that we explicitly set the `isSerial` field on the `tree.ColumnTableDef` to false in `processSerialInColumnDef` and later check that it is false when constructing the column desccriptor in `MakeColumnDefDescs`. Seems to me this is a check to ensure that we've handled `SERIAL` properly. Assuming this is the only reason, and we want to continue having this safeguard, I'd suggest adding another field to `ColumnTableDef` that is set to true once we've handled "SERIAL" by creating the requisite sequence. Then we can keep the check and also save SERIAL information on the column descriptor for this builtin. 

There's still the backward compatibility issue here though for column descriptors that don't have the serial information saved on them. 
___
*[pkg/sql/sequence.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52436#-ME3KFPc6yVoM4nmsSFK:-ME3KFPc6yVoM4nmsSFL:bk8dbr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/018a7a2fddbc00fda40d995e12cfa0965f708e03/pkg/sql/sequence.go#L60)):*
> ```Go
> 				}
> 				// Ensure they are part of the same catalog and schema.
> 				if seq.GetParentID() == tableDesc.GetParentID() && seq.GetParentSchemaID() == tableDesc.GetParentSchemaID() {
> ```

This doesn't always hold even if it is a true SERIAL column, see #51090. Do you still want this check?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52436)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @dt)

___
*[pkg/sql/sequence.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52436#-ME3Fvno9tmZ789AepMF:-ME3Q5se3eAs5pLaAJOr:bdwttvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/018a7a2fddbc00fda40d995e12cfa0965f708e03/pkg/sql/sequence.go#L49)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Can we save this information on the column descriptor during `processSerialInColumnDef`? I see that we explicitly set the `isSerial` field on the `tree.ColumnTableDef` to false in `processSerialInColumnDef` and later check that it is false when constructing the column desccriptor in `MakeColumnDefDescs`. Seems to me this is a check to ensure that we've handled `SERIAL` properly. Assuming this is the only reason, and we want to continue having this safeguard, I'd suggest adding another field to `ColumnTableDef` that is set to true once we've handled "SERIAL" by creating the requisite sequence. Then we can keep the check and also save SERIAL information on the column descriptor for this builtin. 

There's still the backward compatibility issue here though for column descriptors that don't have the serial information saved on them. 
</blockquote></details>

We could, bit nervous about backwards compat and also how this looks in backup and restore. If we emit SERIAL, we'd have to always set sql_sequence to true when creating a table which is a large flow on effect. If we don't emit SERIAL even though we saved the definition, it'll look funny afterwards...
___
*[pkg/sql/sequence.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52436#-ME3KFPc6yVoM4nmsSFK:-ME3Q4olB1hbGVn7U4hR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/018a7a2fddbc00fda40d995e12cfa0965f708e03/pkg/sql/sequence.go#L60)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This doesn't always hold even if it is a true SERIAL column, see #51090. Do you still want this check?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/sequence.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52436#-ME3JuAM-Ri2WAHz58X8:-ME3QbV25YLB77_gExj-:b-yrje87) ([raw file](https://github.com/cockroachdb/cockroach/blob/018a7a2fddbc00fda40d995e12cfa0965f708e03/pkg/sql/sequence.go#L48)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: s/used/uses/g
</blockquote></details>

not sure that's more english


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
looks good, modulo issue + comment nit. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52436#-:-ME3n1d02H7hOpObtrsY:b-8o4jp6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @dt, and @otan)

___
*[pkg/sql/sequence.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52436#-ME3JuAM-Ri2WAHz58X8:-ME3kjGuC-Q4eYB7AMpP:b8lixuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/018a7a2fddbc00fda40d995e12cfa0965f708e03/pkg/sql/sequence.go#L48)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

not sure that's more english
</blockquote></details>

my bad, I guess I meant change the sentence to read "Guess by assuming it is SERIAL if it only uses 1 sequence". The column is using the sequence, not the other way around. 
___
*[pkg/sql/sequence.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52436#-ME3Fvno9tmZ789AepMF:-ME3m2ls7fsQFynuJvWs:bac32xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/018a7a2fddbc00fda40d995e12cfa0965f708e03/pkg/sql/sequence.go#L49)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

We could, bit nervous about backwards compat and also how this looks in backup and restore. If we emit SERIAL, we'd have to always set sql_sequence to true when creating a table which is a large flow on effect. If we don't emit SERIAL even though we saved the definition, it'll look funny afterwards...
</blockquote></details>

Hmm I'm not sure about implications on backup/restore. I'm fine if we file an issue to track how this differs from PG and linking it here. 


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52436)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @dt)

___
*[pkg/sql/sequence.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52436#-ME3Fvno9tmZ789AepMF:-ME3nY_J116XPGlABTiI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/018a7a2fddbc00fda40d995e12cfa0965f708e03/pkg/sql/sequence.go#L49)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Hmm I'm not sure about implications on backup/restore. I'm fine if we file an issue to track how this differs from PG and linking it here. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
builtins: implement pg_get_serial_sequence

Encountered this one whilst importing geospatial data - we need to
ability to use pg_get_serial_sequence for IMPORTing certain dumps
for ogr2ogr.

We have no notion of saving "SERIAL" at the moment, so make some
heuristics to calculate this. Let me know if we should be doing
something else.

Release note (sql change): Add the pg_get_serial_sequence builtin.

otan commented:
bors r=arulajmani
